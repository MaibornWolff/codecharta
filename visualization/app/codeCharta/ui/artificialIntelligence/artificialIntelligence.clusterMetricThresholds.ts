export const clusterMetricThresholds = {
	java: {
		// Thresholds derived without any pre-clustering from benchmark data
		notClustered: {
			mcc: {
				percentile70: 58,
				percentile80: 90,
				percentile90: 165,
				percentile95: 271
			},
			rloc: {
				percentile70: 365,
				percentile80: 554,
				percentile90: 1001,
				percentile95: 1665
			},
			loc: {
				percentile70: 365,
				percentile80: 554,
				percentile90: 1001,
				percentile95: 1665
			},
			cognitive_complexity: {
				percentile70: 30,
				percentile80: 56,
				percentile90: 124,
				percentile95: 226
			},
			code_smell: {
				percentile70: 5,
				percentile80: 9,
				percentile90: 22,
				percentile95: 50
			},
			comment_lines: {
				percentile70: 35,
				percentile80: 62,
				percentile90: 135,
				percentile95: 273
			},
			max_nesting_level: {
				percentile70: 2,
				percentile80: 3,
				percentile90: 4,
				percentile95: 5
			},
			functions: {
				percentile70: 29,
				percentile80: 44,
				percentile90: 75,
				percentile95: 126
			},
			statements: {
				percentile70: 157,
				percentile80: 245,
				percentile90: 439,
				percentile95: 727
			}
		}
	}
}
