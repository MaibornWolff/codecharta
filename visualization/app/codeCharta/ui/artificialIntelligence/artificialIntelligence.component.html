<md-menu class="artificial-intelligence-component">
	<md-button class="ai-button" ng-click="$mdMenu.open($event);" title="Open AI Panel">
		<i class="fa fa-cogs"></i>
	</md-button>

	<md-menu-content class="ai-drop-down" width="7">
		<md-menu-item class="ai-item">Nothing to see here yet</md-menu-item>
	</md-menu-content>
</md-menu>
