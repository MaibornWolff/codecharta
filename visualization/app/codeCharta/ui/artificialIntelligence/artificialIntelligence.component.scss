.artificial-intelligence-component {
	.suspicious-metrics-button,
	.high-risk-profile-button {
		height: 22px;
		min-height: 0;
		line-height: 23px;
		min-width: 45px;
		margin: 0;
		padding: 0;
		font-size: 9pt;
		text-transform: none;
		background-color: rgba(230, 230, 230, 0.96);
		overflow: visible;

		&.plus {
			min-width: 30px;
		}

		.ai-badge {
			position: absolute;
			top: 12px;
			padding: 4px 6px;
			border-radius: 50%;
			background: red;
			color: white;
			font-size: 10px;
			width: 18px;
			height: 18px;
			line-height: 10px;
		}

		&:hover {
			background-color: #c8c8c8;
		}
	}
}
.ai-drop-down {
	padding: 16px;
	width: fit-content;

	md-menu-item {
		height: auto;
		min-height: auto;
	}

	md-menu-item:not(:last-child) {
		margin-bottom: 16px;
	}

	&.metrics {
		.title span {
			padding: 0;
			margin: 0;
			font-weight: bold;
		}

		.suspicious-metrics {
			margin: 0;
			padding: 0;
			list-style: none;
			width: max-content;
		}

		.suspicious-metric {
			.metric-button {
				border-radius: 5px;
				background-color: rgba(230, 230, 230, 0.96);
				padding: 0 12px;
				text-transform: none;
				margin: 0;

				&.high-risk {
					border-radius: 5px 0 0 5px;
				}
			}

			.risk-button {
				border-radius: 0 5px 5px 0;
				background-color: rgba(230, 230, 230, 0.96);
				margin: 0;
				min-width: 36px;
				height: 36px;
				color: #9c0e6a;
			}

			.risk-button:hover,
			.metric-button:hover {
				background-color: #c8c8c8;
			}

			&:not(:last-child) {
				margin-bottom: 12px;
			}
		}
	}

	&.summary {
		.title {
			flex-direction: column;
			gap: 6px;
			font-weight: bold;

			p {
				padding: 0;
				margin: 0;
			}
		}

		.risk-profiles {
			.risk-profile {
				height: 36px;
				text-align: center;
				display: flex;
				justify-content: center;
				align-items: center;
				padding: 0;
				overflow: hidden;

				&.low-risk,
				&.moderate-risk,
				&.high-risk,
				&.very-high-risk {
					color: black;
				}
			}

			.low-risk {
				background-color: #5cb34d;
			}

			.moderate-risk {
				background-color: #eef432;
			}

			.high-risk {
				background-color: #cc2900;
			}

			.very-high-risk {
				background-color: #9c0e6a;
			}
		}
	}
}
