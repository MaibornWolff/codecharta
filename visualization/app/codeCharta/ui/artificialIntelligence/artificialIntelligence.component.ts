import "./artificialIntelligence.component.scss"
import debounce from "lodash.debounce"
import { FilesSelectionSubscriber, FilesService } from "../../state/store/files/files.service"
import { FileState } from "../../model/files/files"
import { IRootScopeService } from "angular"
import { buildCustomConfigFromState } from "../../util/customConfigBuilder"
import { CustomConfigHelper } from "../../util/customConfigHelper"
import { StoreService } from "../../state/store.service"
import { klona } from "klona"
import { CustomConfig, CustomConfigMapSelectionMode } from "../../model/customConfig/customConfig.api.model"
import { pushSorted } from "../../util/nodeDecorator"
import { BlacklistItem, BlacklistType, CodeMapNode, ColorRange, NodeType, State } from "../../codeCharta.model"
import { hierarchy } from "d3-hierarchy"
import { getVisibleFileStates } from "../../model/files/files.helper"
import { metricThresholds } from "./artificialIntelligence.metricThresholds"
import { defaultMapColors } from "../../state/store/appSettings/mapColors/mapColors.actions"
import { ThreeOrbitControlsService } from "../codeMap/threeViewer/threeOrbitControlsService"
import { ThreeCameraService } from "../codeMap/threeViewer/threeCameraService"
import { BlacklistService, BlacklistSubscriber } from "../../state/store/fileSettings/blacklist/blacklist.service"
import { isPathBlacklisted } from "../../util/codeMapHelper"

interface MetricValues {
	[metric: string]: number[]
}

interface MetricAssessmentResults {
	suspiciousMetrics: Map<string, ColorRange>
	unsuspiciousMetrics: string[]
	outliersThresholds: Map<string, number>
}

interface MetricSuggestionParameters {
	metric: string
	from: number
	to: number
	generalCustomConfigId: string
	outlierCustomConfigId?: string
}

export class ArtificialIntelligenceController implements FilesSelectionSubscriber, BlacklistSubscriber {
	private _viewModel: {
		analyzedProgrammingLanguage: string
		suspiciousMetricSuggestionLinks: MetricSuggestionParameters[]
		unsuspiciousMetrics: string[]
		riskProfile: {
			lowRisk: number
			moderateRisk: number
			highRisk: number
			veryHighRisk: number
		}
	} = {
		analyzedProgrammingLanguage: undefined,
		suspiciousMetricSuggestionLinks: [],
		unsuspiciousMetrics: [],
		riskProfile: {
			lowRisk: 0,
			moderateRisk: 0,
			highRisk: 0,
			veryHighRisk: 0
		}
	}

	private debounceCalculation: () => void
	private fileState: FileState
	private blacklist: BlacklistItem[] = []

	constructor(
		private $rootScope: IRootScopeService,
		private storeService: StoreService,
		private threeOrbitControlsService: ThreeOrbitControlsService,
		private threeCameraService: ThreeCameraService
	) {
		"ngInject"
		FilesService.subscribe(this.$rootScope, this)
		BlacklistService.subscribe(this.$rootScope, this)

		this.debounceCalculation = debounce(() => {
			this.calculate()
		}, 10)
	}

	applyCustomConfig(configId: string) {
		CustomConfigHelper.applyCustomConfig(configId, this.storeService, this.threeCameraService, this.threeOrbitControlsService)
	}

	onBlacklistChanged(blacklist: BlacklistItem[]) {
		this.blacklist = blacklist
		this.fileState = getVisibleFileStates(this.storeService.getState().files)[0]

		if (this.fileState !== undefined) {
			this.debounceCalculation()
		}
	}

	onFilesSelectionChanged(files: FileState[]) {
		const fileState = getVisibleFileStates(files)[0]
		if (fileState === undefined) {
			return
		}

		this.fileState = fileState
		this.debounceCalculation()
	}

	private calculate() {
		const mainProgrammingLanguage = this.getMostFrequentLanguage(this.fileState.file.map)
		this._viewModel.analyzedProgrammingLanguage = mainProgrammingLanguage

		this.clearRiskProfile()

		if (mainProgrammingLanguage !== undefined) {
			this.calculateRiskProfile(this.fileState, mainProgrammingLanguage, "mcc")
			this.createCustomConfigSuggestions(this.fileState, mainProgrammingLanguage)
		}
	}

	private clearRiskProfile() {
		this._viewModel.riskProfile = undefined
	}

	private calculateRiskProfile(fileState: FileState, programmingLanguage, metricName) {
		let totalRloc = 0
		let numberOfRlocLowRisk = 0
		let numberOfRlocModerateRisk = 0
		let numberOfRlocHighRisk = 0
		let numberOfRlocVeryHighRisk = 0

		const languageSpecificThresholds = this.getAssociatedMetricThresholds(programmingLanguage)
		const thresholds = languageSpecificThresholds[metricName]

		for (const { data } of hierarchy(fileState.file.map)) {
			// TODO calculate risk profile only for focused or currently visible but not excluded files.
			if (
				data.type !== NodeType.FILE ||
				isPathBlacklisted(data.path, this.blacklist, BlacklistType.exclude) ||
				data.attributes[metricName] === undefined ||
				data.attributes["rloc"] === undefined ||
				this.getFileExtension(data.name) !== programmingLanguage
			) {
				continue
			}

			const nodeMetricValue = data.attributes[metricName]
			const nodeRlocValue = data.attributes["rloc"]

			totalRloc += nodeRlocValue

			// Idea: We could calculate risk profiles per directory in the future.
			if (nodeMetricValue <= thresholds.percentile70) {
				numberOfRlocLowRisk += nodeRlocValue
			} else if (nodeMetricValue <= thresholds.percentile80) {
				numberOfRlocModerateRisk += nodeRlocValue
			} else if (nodeMetricValue <= thresholds.percentile90) {
				numberOfRlocHighRisk += nodeRlocValue
			} else {
				numberOfRlocVeryHighRisk += nodeRlocValue
			}
		}

		if (totalRloc === 0) {
			return
		}

		this._viewModel.riskProfile = {
			lowRisk: Math.ceil((numberOfRlocLowRisk / totalRloc) * 100),
			moderateRisk: Math.ceil((numberOfRlocModerateRisk / totalRloc) * 100),
			highRisk: Math.ceil((numberOfRlocHighRisk / totalRloc) * 100),
			veryHighRisk: Math.ceil((numberOfRlocVeryHighRisk / totalRloc) * 100)
		}
	}

	private createCustomConfigSuggestions(fileState: FileState, programmingLanguage) {
		const metricValues = this.getSortedMetricValues(fileState, programmingLanguage)
		const metricAssessmentResults = this.findGoodAndBadMetrics(metricValues, programmingLanguage)

		const noticeableMetricSuggestionLinks = new Map<string, MetricSuggestionParameters>()
		const newCustomConfigs: CustomConfig[] = []

		for (const [metricName, colorRange] of metricAssessmentResults.suspiciousMetrics) {
			const suspiciousMetricConfig = this.createOrUpdateCustomConfig(
				`Suspicious Files - Metric ${metricName.toUpperCase()} - (AI)`,
				this.prepareOverviewConfigState(metricName, colorRange.from, colorRange.to),
				fileState
			)

			newCustomConfigs.push(suspiciousMetricConfig)

			noticeableMetricSuggestionLinks.set(metricName, {
				metric: metricName,
				...colorRange,
				generalCustomConfigId: suspiciousMetricConfig.id
			})

			const outlierThreshold = metricAssessmentResults.outliersThresholds.get(metricName)
			if (outlierThreshold > 0) {
				const highRiskMetricConfig = this.createOrUpdateCustomConfig(
					`Very High Risk Files - Metric ${metricName.toUpperCase()} - (AI)`,
					this.prepareOutlierConfigState(metricName, outlierThreshold - 1, outlierThreshold),
					fileState
				)

				newCustomConfigs.push(highRiskMetricConfig)

				noticeableMetricSuggestionLinks.get(metricName).outlierCustomConfigId = highRiskMetricConfig.id
			}
		}

		CustomConfigHelper.addCustomConfigs(newCustomConfigs)

		this._viewModel.suspiciousMetricSuggestionLinks = [...noticeableMetricSuggestionLinks.values()]
		this._viewModel.unsuspiciousMetrics = metricAssessmentResults.unsuspiciousMetrics
	}

	private findGoodAndBadMetrics(metricValues, programmingLanguage): MetricAssessmentResults {
		const metricAssessmentResults: MetricAssessmentResults = {
			suspiciousMetrics: new Map<string, ColorRange>(),
			unsuspiciousMetrics: [],
			outliersThresholds: new Map<string, number>()
		}

		const languageSpecificMetricThresholds = this.getAssociatedMetricThresholds(programmingLanguage)

		for (const metricName of Object.keys(languageSpecificMetricThresholds)) {
			const valuesOfMetric = metricValues[metricName]
			if (valuesOfMetric === undefined) {
				continue
			}

			const thresholdConfig = languageSpecificMetricThresholds[metricName]
			const maxMetricValue = Math.max(...valuesOfMetric)

			if (maxMetricValue <= thresholdConfig.percentile70) {
				metricAssessmentResults.unsuspiciousMetrics.push(metricName)
			} else if (maxMetricValue > thresholdConfig.percentile70) {
				metricAssessmentResults.suspiciousMetrics.set(metricName, {
					from: thresholdConfig.percentile70,
					to: thresholdConfig.percentile80,
					max: 0,
					min: 0
				})

				if (maxMetricValue > thresholdConfig.percentile90) {
					metricAssessmentResults.outliersThresholds.set(metricName, thresholdConfig.percentile90)
				}
			}
		}

		return metricAssessmentResults
	}

	private createOrUpdateCustomConfig(configName: string, state: State, fileState: FileState) {
		const customConfig = CustomConfigHelper.getCustomConfigByName(
			CustomConfigMapSelectionMode.SINGLE,
			[fileState.file.fileMeta.fileName],
			configName
		)

		// If it exists, create a fresh one with current thresholds
		if (customConfig !== null) {
			CustomConfigHelper.deleteCustomConfig(customConfig.id)
		}

		return buildCustomConfigFromState(configName, state)
	}

	private prepareOverviewConfigState(metricName: string, colorRangeFrom: number, colorRangeTo: number) {
		const state = klona(this.storeService.getState())

		// just use rloc as area metric
		state.dynamicSettings.areaMetric = "rloc"

		// use bad metric as height and color metric
		state.dynamicSettings.heightMetric = metricName
		state.dynamicSettings.colorMetric = metricName
		//state.dynamicSettings.colorRange.from = bugOutlier.valueOf() / 2
		//state.dynamicSettings.colorRange.to = bugOutlier.valueOf()
		state.dynamicSettings.colorRange.from = colorRangeFrom
		state.dynamicSettings.colorRange.to = colorRangeTo

		state.appSettings.mapColors.positive = defaultMapColors.positive
		state.appSettings.mapColors.neutral = defaultMapColors.neutral
		state.appSettings.mapColors.negative = defaultMapColors.negative

		return state
	}

	private prepareOutlierConfigState(metricName: string, colorRangeFrom: number, colorRangeTo: number) {
		const state = klona(this.storeService.getState())

		// just use rloc as area metric
		state.dynamicSettings.areaMetric = "rloc"

		// use bad metric as height and color metric
		state.dynamicSettings.heightMetric = metricName
		state.dynamicSettings.colorMetric = metricName

		state.dynamicSettings.colorRange.to = colorRangeTo
		state.dynamicSettings.colorRange.from = colorRangeFrom

		state.appSettings.mapColors.positive = "#ffffff"
		state.appSettings.mapColors.neutral = "#ffffff"
		state.appSettings.mapColors.negative = "#A900C0"

		return state
	}

	private getSortedMetricValues(fileState: FileState, programmingLanguage): MetricValues {
		const metricValues: MetricValues = {}

		for (const { data } of hierarchy(fileState.file.map)) {
			if (
				data.type !== NodeType.FILE ||
				isPathBlacklisted(data.path, this.blacklist, BlacklistType.exclude) ||
				this.getFileExtension(data.name) !== programmingLanguage
			) {
				continue
			}

			for (const metricIndex of Object.keys(data.attributes)) {
				const value = data.attributes[metricIndex]
				if (value > 0) {
					if (metricValues[metricIndex] === undefined) {
						metricValues[metricIndex] = []
					}
					pushSorted(metricValues[metricIndex], data.attributes[metricIndex])
				}
			}
		}

		return metricValues
	}

	private getFileExtension(fileName: string) {
		return fileName.includes(".") ? fileName.slice(fileName.lastIndexOf(".") + 1) : undefined
	}

	private getMostFrequentLanguage(map: CodeMapNode) {
		const numberOfFilesPerLanguage = []

		for (const { data } of hierarchy(map)) {
			if (!data.name.includes(".")) {
				continue
			}

			if (data.type === NodeType.FILE) {
				const fileExtension = data.name.slice(data.name.lastIndexOf(".") + 1)
				numberOfFilesPerLanguage[fileExtension] = numberOfFilesPerLanguage[fileExtension] + 1 || 1
			}
		}

		if (Object.keys(numberOfFilesPerLanguage).length === 0) {
			return
		}

		return Object.keys(numberOfFilesPerLanguage).reduce((a, b) => {
			return numberOfFilesPerLanguage[a] > numberOfFilesPerLanguage[b] ? a : b
		})
	}

	private getAssociatedMetricThresholds(programmingLanguage) {
		return programmingLanguage === "java" ? metricThresholds["java"] : metricThresholds["miscellaneous"]
	}
}

export const artificialIntelligenceComponent = {
	selector: "artificialIntelligenceComponent",
	template: require("./artificialIntelligence.component.html"),
	controller: ArtificialIntelligenceController
}
