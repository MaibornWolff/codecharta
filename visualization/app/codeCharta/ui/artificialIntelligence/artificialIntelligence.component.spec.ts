import "./artificialIntelligence.module"
import { instantiateModule, getService } from "../../../../mocks/ng.mockhelper"
import { IRootScopeService } from "angular"
import { StoreService } from "../../state/store.service"
import { ArtificialIntelligenceController } from "./artificialIntelligence.component"
import { FilesService } from "../../state/store/files/files.service"
import { setFiles } from "../../state/store/files/files.actions"
import { FILE_STATES, FILE_STATES_JAVA, FILE_STATES_UNSELECTED, STATE } from "../../util/dataMocks"
import { klona } from "klona"
import { BlacklistType, ColorRange } from "../../codeCharta.model"
import { BlacklistService } from "../../state/store/fileSettings/blacklist/blacklist.service"
import { ExperimentalFeaturesEnabledService } from "../../state/store/appSettings/enableExperimentalFeatures/experimentalFeaturesEnabled.service"
import { setExperimentalFeaturesEnabled } from "../../state/store/appSettings/enableExperimentalFeatures/experimentalFeaturesEnabled.actions"
import { setColorRange } from "../../state/store/dynamicSettings/colorRange/colorRange.actions"
import { setColorMetric } from "../../state/store/dynamicSettings/colorMetric/colorMetric.actions"
import { setHeightMetric } from "../../state/store/dynamicSettings/heightMetric/heightMetric.actions"
import { setAreaMetric } from "../../state/store/dynamicSettings/areaMetric/areaMetric.actions"
import { setMapColors } from "../../state/store/appSettings/mapColors/mapColors.actions"
import { setState } from "../../state/store/state.actions"

describe("ArtificialIntelligenceController", () => {
	let artificialIntelligenceController: ArtificialIntelligenceController
	let $rootScope: IRootScopeService
	let storeService: StoreService

	beforeEach(() => {
		restartSystem()
		rebuildController()
	})

	function restartSystem() {
		instantiateModule("app.codeCharta.ui.artificialIntelligence")

		$rootScope = getService<IRootScopeService>("$rootScope")
		storeService = getService<StoreService>("storeService")

		storeService.dispatch(setFiles(FILE_STATES))
	}

	function rebuildController() {
		artificialIntelligenceController = new ArtificialIntelligenceController($rootScope, storeService)

		// Overwrite debounce with original function, otherwise calculate() will not be called
		artificialIntelligenceController["debounceCalculation"] = artificialIntelligenceController["calculate"]
	}

	describe("constructor", () => {
		it("should subscribe to file service", () => {
			FilesService.subscribe = jest.fn()

			rebuildController()

			expect(FilesService.subscribe).toHaveBeenCalledWith($rootScope, artificialIntelligenceController)
		})

		it("should subscribe to blacklist service", () => {
			BlacklistService.subscribe = jest.fn()

			rebuildController()

			expect(BlacklistService.subscribe).toHaveBeenCalledWith($rootScope, artificialIntelligenceController)
		})

		it("should subscribe to experimentalFeaturesEnabled service", () => {
			ExperimentalFeaturesEnabledService.subscribe = jest.fn()

			rebuildController()

			expect(ExperimentalFeaturesEnabledService.subscribe).toHaveBeenCalledWith($rootScope, artificialIntelligenceController)
		})
	})

	describe("calculations", () => {
		it("should not calculate suspicious metrics when experimental features are disabled", function () {
			artificialIntelligenceController["getMostFrequentLanguage"] = jest.fn()
			artificialIntelligenceController["clearRiskProfile"] = jest.fn()
			artificialIntelligenceController["calculateRiskProfile"] = jest.fn()
			artificialIntelligenceController["calculateSuspiciousMetrics"] = jest.fn()
			artificialIntelligenceController["fileState"] = FILE_STATES_JAVA[0]

			storeService.dispatch(setExperimentalFeaturesEnabled(false))
			artificialIntelligenceController.onExperimentalFeaturesEnabledChanged(false)

			expect(artificialIntelligenceController["getMostFrequentLanguage"]).not.toHaveBeenCalled()
			expect(artificialIntelligenceController["clearRiskProfile"]).not.toHaveBeenCalled()
			expect(artificialIntelligenceController["calculateRiskProfile"]).not.toHaveBeenCalled()
			expect(artificialIntelligenceController["calculateSuspiciousMetrics"]).not.toHaveBeenCalled()
		})

		it("should calculate suspicious metrics when experimental features are enabled", function () {
			artificialIntelligenceController["getMostFrequentLanguage"] = jest.fn().mockReturnValue("java")
			artificialIntelligenceController["clearRiskProfile"] = jest.fn()
			artificialIntelligenceController["calculateRiskProfile"] = jest.fn()
			artificialIntelligenceController["calculateSuspiciousMetrics"] = jest.fn()
			artificialIntelligenceController["fileState"] = FILE_STATES_JAVA[0]

			storeService.dispatch(setExperimentalFeaturesEnabled(true))
			artificialIntelligenceController.onExperimentalFeaturesEnabledChanged(true)

			expect(artificialIntelligenceController["getMostFrequentLanguage"]).toHaveBeenCalled()
			expect(artificialIntelligenceController["_viewModel"].analyzedProgrammingLanguage).toBe("java")
			expect(artificialIntelligenceController["clearRiskProfile"]).toHaveBeenCalled()
			expect(artificialIntelligenceController["calculateRiskProfile"]).toHaveBeenCalled()
			expect(artificialIntelligenceController["calculateSuspiciousMetrics"]).toHaveBeenCalled()
		})
	})

	describe("apply suspicious metric", () => {
		it("should call store.dispatch with all needed states for desired suspicious metric view", () => {
			storeService.dispatch(setState(STATE))
			const testMetricSuggestionParameters = {
				metric: "loc",
				from: 365,
				to: 554,
				max: 0,
				min: 0,
				isOutlier: undefined
			}

			const colorRange: ColorRange = {
				from: testMetricSuggestionParameters.from,
				to: testMetricSuggestionParameters.to,
				max: 0,
				min: 0
			}

			artificialIntelligenceController["_viewModel"].suspiciousMetricSuggestionLinks = [testMetricSuggestionParameters]
			storeService.dispatch = jest.fn()
			artificialIntelligenceController.applySuspiciousMetric(testMetricSuggestionParameters, false)

			expect(storeService.dispatch).toHaveBeenCalledWith(setAreaMetric("rloc"))
			expect(storeService.dispatch).toHaveBeenCalledWith(setHeightMetric("loc"))
			expect(storeService.dispatch).toHaveBeenCalledWith(setColorMetric("loc"))
			expect(storeService.dispatch).toHaveBeenCalledWith(setColorRange(colorRange))
			expect(storeService.dispatch).toHaveBeenCalledWith(setMapColors(STATE.appSettings.mapColors))
		})

		it("should call store.dispatch with all needed states for desired very high risk metric view", () => {
			storeService.dispatch(setState(STATE))
			const mapColors = { ...STATE.appSettings.mapColors }
			mapColors.positive = "#ffffff"
			mapColors.neutral = "#ffffff"
			mapColors.negative = "#A900C0"

			const testMetricSuggestionParameters = {
				metric: "loc",
				from: 365,
				to: 554,
				max: 0,
				min: 0,
				isOutlier: true
			}

			const colorRange: ColorRange = {
				from: testMetricSuggestionParameters.from,
				to: testMetricSuggestionParameters.to,
				max: 0,
				min: 0
			}

			artificialIntelligenceController["_viewModel"].suspiciousMetricSuggestionLinks = [testMetricSuggestionParameters]
			storeService.dispatch = jest.fn()
			artificialIntelligenceController.applySuspiciousMetric(testMetricSuggestionParameters, true)

			expect(storeService.dispatch).toHaveBeenCalledWith(setAreaMetric("rloc"))
			expect(storeService.dispatch).toHaveBeenCalledWith(setHeightMetric("loc"))
			expect(storeService.dispatch).toHaveBeenCalledWith(setColorMetric("loc"))
			expect(storeService.dispatch).toHaveBeenCalledWith(setColorRange(colorRange))
			expect(storeService.dispatch).toHaveBeenCalledWith(setMapColors(mapColors))
		})
	})

	describe("on files selection changed", () => {
		it("should do nothing if no file is selected", () => {
			artificialIntelligenceController["getMostFrequentLanguage"] = jest.fn()
			artificialIntelligenceController.onFilesSelectionChanged(FILE_STATES_UNSELECTED)
			expect(artificialIntelligenceController["getMostFrequentLanguage"]).not.toHaveBeenCalled()
		})

		it("should not calculate risk profile and suspicious metrics on empty main programming language", () => {
			artificialIntelligenceController["clearRiskProfile"] = jest.fn()
			artificialIntelligenceController["calculateRiskProfile"] = jest.fn()
			artificialIntelligenceController["createCustomConfigSuggestions"] = jest.fn()
			storeService.dispatch(setExperimentalFeaturesEnabled(true))
			artificialIntelligenceController.onFilesSelectionChanged(FILE_STATES)

			expect(artificialIntelligenceController["clearRiskProfile"]).toHaveBeenCalled()
			expect(artificialIntelligenceController["calculateRiskProfile"]).not.toHaveBeenCalled()
			expect(artificialIntelligenceController["createCustomConfigSuggestions"]).not.toHaveBeenCalled()
		})

		it("should clear and calculate risk profile for Java map", () => {
			artificialIntelligenceController["clearRiskProfile"] = jest.fn()
			artificialIntelligenceController["calculateSuspiciousMetrics"] = jest.fn()

			storeService.dispatch(setExperimentalFeaturesEnabled(true))
			artificialIntelligenceController.onFilesSelectionChanged(FILE_STATES_JAVA)

			expect(artificialIntelligenceController["_viewModel"].analyzedProgrammingLanguage).toBe("java")
			expect(artificialIntelligenceController["clearRiskProfile"]).toHaveBeenCalled()
			expect(artificialIntelligenceController["_viewModel"].riskProfile).toMatchSnapshot()
			expect(artificialIntelligenceController["calculateSuspiciousMetrics"]).toHaveBeenCalled()
		})

		it("should calculate risk profile and should not exceed 100 percent", () => {
			storeService.dispatch(setExperimentalFeaturesEnabled(true))
			artificialIntelligenceController.onFilesSelectionChanged(FILE_STATES_JAVA)

			const sumOfRiskPercentage = Object.values(artificialIntelligenceController["_viewModel"].riskProfile).reduce((a, b) => a + b)
			const expectedRiskProfile = {
				highRisk: 37,
				lowRisk: 46,
				moderateRisk: 17,
				veryHighRisk: 0
			}

			expect(artificialIntelligenceController["_viewModel"].riskProfile).toEqual(expectedRiskProfile)
			expect(sumOfRiskPercentage).toEqual(100)
		})

		it("should calculate suspicious metrics sorted by isOutlier", () => {
			artificialIntelligenceController["clearRiskProfile"] = jest.fn()
			artificialIntelligenceController["calculateRiskProfile"] = jest.fn()

			storeService.dispatch(setExperimentalFeaturesEnabled(true))
			artificialIntelligenceController.onFilesSelectionChanged(FILE_STATES_JAVA)

			expect(artificialIntelligenceController["_viewModel"].analyzedProgrammingLanguage).toBe("java")
			expect(artificialIntelligenceController["clearRiskProfile"]).toHaveBeenCalled()
			expect(artificialIntelligenceController["calculateRiskProfile"]).toHaveBeenCalled()

			expect(artificialIntelligenceController["_viewModel"].suspiciousMetricSuggestionLinks).toMatchSnapshot()
			expect(artificialIntelligenceController["_viewModel"].unsuspiciousMetrics).toMatchSnapshot()
		})

		it("should calculate risk profile for not excluded files only", () => {
			artificialIntelligenceController["calculateSuspiciousMetrics"] = jest.fn()

			const codeMapNodeToExclude = FILE_STATES_JAVA[0].file.map.children[0].children[0]

			artificialIntelligenceController["blacklist"] = [
				{
					path: codeMapNodeToExclude.path,
					type: BlacklistType.exclude
				}
			]

			storeService.dispatch(setExperimentalFeaturesEnabled(true))
			artificialIntelligenceController.onFilesSelectionChanged(FILE_STATES_JAVA)

			expect(artificialIntelligenceController["_viewModel"].analyzedProgrammingLanguage).toBe("java")
			expect(artificialIntelligenceController["_viewModel"].riskProfile).toMatchSnapshot()
			expect(artificialIntelligenceController["calculateSuspiciousMetrics"]).toHaveBeenCalled()
		})

		it("should calculate risk profile but skip files with missing metrics", () => {
			artificialIntelligenceController["calculateSuspiciousMetrics"] = jest.fn()

			const FILE_STATES_MISSING_METRICS = klona(FILE_STATES_JAVA)
			for (const codeMapNode of FILE_STATES_MISSING_METRICS[0].file.map.children) {
				codeMapNode.children.map(childCodeMapNode => {
					childCodeMapNode.attributes = {}
				})
			}

			storeService.dispatch(setExperimentalFeaturesEnabled(true))
			artificialIntelligenceController.onFilesSelectionChanged(FILE_STATES_MISSING_METRICS)

			expect(artificialIntelligenceController["_viewModel"].analyzedProgrammingLanguage).toBe("java")
			expect(artificialIntelligenceController["_viewModel"].riskProfile).toBeUndefined()
			expect(artificialIntelligenceController["calculateSuspiciousMetrics"]).toHaveBeenCalled()
		})

		it("should calculate risk profile and suspicious metrics for maps with other programming languages", () => {
			artificialIntelligenceController["calculateRiskProfile"] = jest.fn()
			artificialIntelligenceController["calculateSuspiciousMetrics"] = jest.fn()

			const FILE_STATES_OTHER = klona(FILE_STATES_JAVA)
			for (const codeMapNode of FILE_STATES_OTHER[0].file.map.children) {
				codeMapNode.children.map(childCodeMapNode => {
					childCodeMapNode.name = childCodeMapNode.name.replace(/\.java/, ".other")
				})
			}

			storeService.dispatch(setExperimentalFeaturesEnabled(true))
			artificialIntelligenceController.onFilesSelectionChanged(FILE_STATES_OTHER)

			expect(artificialIntelligenceController["_viewModel"].analyzedProgrammingLanguage).toBe("other")
			expect(artificialIntelligenceController["calculateRiskProfile"]).toHaveBeenCalled()
			expect(artificialIntelligenceController["calculateSuspiciousMetrics"]).toHaveBeenCalled()
		})
	})

	describe("on blacklist changed", () => {
		it("should do nothing on blacklist change if no file is selected", () => {
			storeService.dispatch(setFiles(FILE_STATES_UNSELECTED))
			artificialIntelligenceController["debounceCalculation"] = jest.fn()

			artificialIntelligenceController.onBlacklistChanged([])
			expect(artificialIntelligenceController["debounceCalculation"]).not.toHaveBeenCalled()
		})

		it("should calculate risk profile on blacklist changed", () => {
			storeService.dispatch(setFiles(FILE_STATES_JAVA))
			artificialIntelligenceController["debounceCalculation"] = jest.fn()

			artificialIntelligenceController.onBlacklistChanged([])
			expect(artificialIntelligenceController["debounceCalculation"]).toHaveBeenCalled()
		})
	})
})
