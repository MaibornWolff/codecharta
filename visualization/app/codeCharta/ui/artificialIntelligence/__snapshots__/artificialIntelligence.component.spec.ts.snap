// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArtificialIntelligenceController on files selection changed should calculate risk profile for not excluded files only 1`] = `
Object {
  "highRisk": 50,
  "lowRisk": 27,
  "moderateRisk": 23,
  "veryHighRisk": 0,
}
`;

exports[`ArtificialIntelligenceController on files selection changed should clear and calculate risk profile for Java map 1`] = `
Object {
  "highRisk": 37,
  "lowRisk": 46,
  "moderateRisk": 17,
  "veryHighRisk": 0,
}
`;

exports[`ArtificialIntelligenceController on files selection changed should create custom config suggestions sorted by outlierCustomConfigId 1`] = `
Array [
  Object {
    "from": 365,
    "generalCustomConfigId": "mocked",
    "max": 0,
    "metric": "loc",
    "min": 0,
    "outlierCustomConfigId": "mocked",
    "to": 554,
  },
  Object {
    "from": 29,
    "generalCustomConfigId": "mocked",
    "max": 0,
    "metric": "functions",
    "min": 0,
    "outlierCustomConfigId": "mocked",
    "to": 44,
  },
  Object {
    "from": 48,
    "generalCustomConfigId": "mocked",
    "max": 0,
    "metric": "mcc",
    "min": 0,
    "to": 71,
  },
]
`;

exports[`ArtificialIntelligenceController on files selection changed should create custom config suggestions sorted by outlierCustomConfigId 2`] = `
Array [
  "rloc (real lines of code)",
]
`;
